### ISSUE TITLE
getGroupMembershipForUser throwing several ECONNREFUSED

### ISSUE THREAD

**Issue by kmz850** [gheeres/node-activedirectory#112]
*getGroupMembershipForUser ERROR for train8: {"code":"ECONNREFUSED","errno":"ECON
NREFUSED","syscall":"connect"}*

I'm getting a steady stream of ECONNREFUSED errors when making calls to getGroupMembershipForUser(). Did some browsing around on the issues here, I've come to conclusion that it's an expensive call and perhaps I shouldn't be using it as I am. employeesArray will contain ~3000 items, but I've cut it down to 300 just for testing and I'm still failing.

Our AD is using referrals from what I understand and I've tried switching ports 389 and 3268 without much luck. Sometimes it will run through without errors... try it again and I get flooded with them.

I've also tried refreshing my AD connection midway through but didn't seem to help, just taking while guesses

    ad = new ActiveDirectory(config);

I notice some of the error responses say "syscall":"write" ... I would expect to only see read and connect?

A code snippet in case you see something obvious.

```javascript
      async.each(employeesArray, function(element, cbEmployeesArray) 
      {

        async.retry({times: 5, interval: 2000}, function(cbGroupMembership, results) 
        {

          ad.getGroupMembershipForUser(element, function(err, groups) 
          {
                if ((err) || (! groups)) {
                  console.log('getGroupMembershipForUser ERROR for ' + element + ': ' + JSON.stringify(err));
                  return cbGroupMembership('error', null); 
                }

                var groupsTemp = new Array();

                _.each(groups, function(groups) {
                      groupsTemp.push(groups.cn);
                });

                var usernameToGroups = {"username": element, "groups": groupsTemp};
                groupMembership.push(usernameToGroups);

                cbGroupMembership(null, 'Success');

          });
        
       }, function(err, result) {
          cbEmployeesArray();
       });

   }, function (err) {
    callback3();
  });
````

**Comment by gheeres** [comment]
Yes. Getting group membership is "expensive" especially if have a bunch of nested groups (i.e. groups that are member of groups, etc.). The Active Directory library will actually go and evaluate (flatten) all of those nested groups. Each group retrieval ends up with another ldap query as they are enumerated. If that group has groups, then those are enumerated as well. So you have an exponential explosion of LDAP queries.

I see that your code is also processing each employee asynchronously which means that the situation is "exponentially exponential". If you have referrals, it's even worse since every referral has to be sent the same request.

In short, the domain controller gets overwhelmed and starts to reject / fail requests.

I'd recommend changing your approach on how you're getting your data (i.e. come at in from a group first perspective) or to only enumerate the groups when they are necessary (i.e. lazy load). How are you using the group information?

**Comment by kmz850** [comment]
This is an inventory type process that will be run maybe once a month or week. The data will be integrated to mind mapping software to better document our systems (ie find a user, tell me what groups they're involved and possible servers and file shares they are accessing... Etc). 

I can probably come at it from a group level first if that method is more efficient. I'll have to see if I can do a scan for all groups and then load all users for that group. Then scrub the data locally. 

Thanks, great tool so far.

**Comment by gheeres** [comment]
You can also see if processing your employees synchronously instead of asynchronously will help. It will be slower, but if this is a batch process... speed shouldn't be as critical.

**Comment by kmz850** [comment]
I'm using async.series to grab employees first then using async.each to loop through each employee to grab their groups. I was under the impression I was doing this synchronously but I'm still a bit new to node so.

**Comment by gheeres** [comment]
No. [async.each](https://github.com/caolan/async#each) runs them in parallel.

**Comment by kmz850** [comment]
Switched to async.eachSeries and the problem cleared right up (at least for now), even when increasing my array size. 

Thank you and keep up the good work.

**Comment by gheeres** [comment]
Thanks. Good luck on you project!
